### PR TITLE
[UIE-141] Organize alerts code and convert to TypeScript

### PR DESCRIPTION
**Pull request by nawatts**
Prefactoring before adding a new alert to prompt users to refresh when Terra UI has been updated.

This moves the Alerts component and service alerts polling into src/alerts and converts them to TypeScript. It also refactors tests a little to better align with current conventions.

## Testing
To manually test, mock the service alerts response.

```js
window.ajaxOverridesStore.set([
  {
    filter: { url: /alerts\.json$/ },
    fn: window.ajaxOverrideUtils.makeSuccess([
      {
        title: 'The systems are down!',
        message: 'Something is terribly wrong'
      }
    ])
  }
])
```

You may also want to change the polling frequency in service-alerts-polling.ts to avoid having to wait a full minute.

### CODE REVIEW

File: packages/components/src/icon.ts [hunk 1]
```diff
@@ -1,14 +1,12 @@
...
 
 import iconLibrary, { IconId } from './icon-library';
 
-export interface IconProps {
-  'aria-label'?: string;
-  'aria-labelledby'?: string;
-  className?: string;
+type SvgProps = JSX.IntrinsicElements['svg'];
+
+export interface IconProps extends SvgProps {
```
Comment (nawatts) [inline packages/components/src/icon.ts:7]: `icon` passes through unknown props. This updates its arguments type accordingly.

File: src/alerts/Alerts.ts [hunk 2]
```diff
@@ -1,18 +1,27 @@
...
+import { Alert as AlertT } from './Alert';
+import { useServiceAlerts } from './service-alerts';
+
+interface AlertProps {
+  alert: AlertT;
+}
+
+const Alert = (props: AlertProps): ReactNode => {
+  const {
+    alert: { title, message, link, linkTitle, severity },
```
Comment (s-rubenstein) [inline src/alerts/Alerts.ts:23]: Non-blocking opt: The pattern here where alert is wrapped in an additional object seems a bit strange to me - should those be lifted up one level?

**Review by s-rubenstein — APPROVED** [review]

File: src/alerts/Alerts.ts (hunk 2 above)
Comment (nawatts) [reply src/alerts/Alerts.ts:23]: Conceptually, the `Alert` component is taking one prop: a single `Alert` object.

The `Alert` component _could_ take all of the `Alert` object's properties as individual props, but I think keeping the alert as a single object makes more sense.

File: src/alerts/Alerts.test.ts [hunk 3]
```diff
@@ -0,0 +1,81 @@
+import { fireEvent, screen, within } from '@testing-library/react';
+import { h } from 'react-hyperscript-helpers';
+import * as Utils from 'src/libs/utils';
+import { asMockedFn, renderWithAppContexts as render } from 'src/testing/test-utils';
+
+import { Alert as AlertT } from './Alert';
```
Comment (msilva-broad) [inline src/alerts/Alerts.test.ts]: as AlertType to be clearer?

File: src/alerts/Alerts.ts [hunk 4]
```diff
@@ -163,7 +183,11 @@ const AlertsIndicator = ({ style }) => {
...
 
     // Have screen readers announce alerts.
-    _.map(({ id, title, message }) => div({ key: id, role: 'alert', className: 'sr-only' }, [div([title]), !!message && div([message])]), alerts),
+    _.map(
+      ({ id, title, message }) =>
+        div({ key: id, role: 'alert', className: 'sr-only' }, [div([title]), !!message && div([message])]),
+      alerts
+    ),
   ]);
 };
```
Comment (msilva-broad) [inline src/alerts/Alerts.ts]: exporting as default is something we are trying to move away from... would it be too disruptive to change the main export to be named instead?

File: src/alerts/Alerts.test.ts (hunk 3 above)
Comment (nawatts) [reply src/alerts/Alerts.test.ts]: Done in https://github.com/DataBiosphere/terra-ui/pull/4400/commits/0958f5e50988c73c9bdcf0f48a0024dadf19c118.

File: src/alerts/Alerts.ts (hunk 4 above)
Comment (nawatts) [reply src/alerts/Alerts.ts]: Done in https://github.com/DataBiosphere/terra-ui/pull/4400/commits/574ab686e3106dfd38cf417ae5107eba364325e1.

File: src/alerts/Alerts.test.ts [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+    jest.restoreAllMocks();
+  });
+
+  it('renders number of alerts', () => {
+    render(h(AlertsIndicator));
+    expect(screen.getByRole('button')).toHaveTextContent(`${testAlerts.length}`);
+  });
+
+  it('renders popup with alerts', () => {
+    render(h(AlertsIndicator));
```
Comment (msilva-broad) [inline src/alerts/Alerts.test.ts:48]: // Arrange, // Act, // Assert section comments on all these tests please.

Comment (nawatts) [reply src/alerts/Alerts.test.ts:48]: Done in https://github.com/DataBiosphere/terra-ui/pull/4400/commits/42f472fa02230a81fac9915b1e827b25e2f85dfe.

File: src/alerts/service-alerts-polling.test.ts [hunk 6]
```diff
@@ -18,9 +24,10 @@ describe('startPollingServiceAlerts', () => {
   });
 
   it('periodically fetches service alerts and updates store', async () => {
```
Comment (msilva-broad) [inline src/alerts/service-alerts-polling.test.ts:26]: // Arrange // Act and // Assert test section comments here and in service-alerts.test.ts too please.

Comment (nawatts) [reply src/alerts/service-alerts-polling.test.ts:26]: Done in e7cbbcfb6183d28ff7c80261f9da72d1d493ef8e.

File: src/alerts/service-alerts.test.ts [hunk 7]
```diff
@@ -19,17 +22,26 @@ afterEach(() => {
...
+    // Act
     await getServiceAlerts();
+
+    // Assert
     expect(mockGetServiceAlerts).toHaveBeenCalled();
   });
 
   it('adds IDs to alerts using hashes of alert content', async () => {
-    Ajax.mockReturnValue({
+    // Assert
```
Comment (msilva-broad) [inline src/alerts/service-alerts.test.ts]: pretty sure you mean // Arrange here

Comment (nawatts) [reply src/alerts/service-alerts.test.ts]: Fixed in https://github.com/DataBiosphere/terra-ui/commit/a8f8baaa2703dc40863bbe95c96ef441828a967d.

**Review by msilva-broad — APPROVED** [review]
great work!

**Pull request merged**